web_external/SpinBox/index.jsx

```diff
@@ -0,0 +1,28 @@
+import React, { Component } from 'react';
+
+import './style.styl';
+
+class SpinBox extends Component {
+    constructor(props) {
+        super(props);
+    }
+
+    calcWidth() {
+        return Math.floor(Math.log10(this.props.max)) + 1;
```

Comment:
The changes are great. 
One thing for this line, I think if the width is decided on the max value we could calculate it on initialize or when the max change, otherwise the calculation needs to be done for every frame change.

Reply:
Do you mean this?

```diff
diff --git a/web_external/SpinBox/index.jsx b/web_external/SpinBox/index.jsx
index 1e79077..a8c0dc9 100644
--- a/web_external/SpinBox/index.jsx
+++ b/web_external/SpinBox/index.jsx
@@ -5,16 +5,13 @@ import './style.styl';
 class SpinBox extends Component {
     constructor(props) {
         super(props);
-    }
-
-    calcWidth() {
-        return Math.floor(Math.log10(this.props.max)) + 1;
+        this.width = Math.floor(Math.log10(this.props.max)) + 1;
     }
     render() {
         return <div className={'spinbox' + (this.props.disabled ? ' disabled' : '')}>
             <input
                 type='number'
-                style={{width: this.calcWidth() + 'em'}}
+                style={{width: this.width + 'em'}}
                 min={this.props.min}
                 max={this.props.max}
                 value={this.props.value}
```

Reply:
Yes. or put it inside the react [componentwillreceiveprops](https://reactjs.org/docs/react-component.html#componentwillreceiveprops) so when a new clip is loaded, the width would adjust accordingly.

Reply:
```diff
diff --git a/web_external/SpinBox/index.jsx b/web_external/SpinBox/index.jsx
index 1e79077..fe1c693 100644
--- a/web_external/SpinBox/index.jsx
+++ b/web_external/SpinBox/index.jsx
@@ -5,16 +5,19 @@ import './style.styl';
 class SpinBox extends Component {
     constructor(props) {
         super(props);
+        this.state = {
+          width: this.calcWidth(props.max)
+        };
     }
-
-    calcWidth() {
+    componentWillReceiveProps(nextProps) {
+        var w = this.calcWidth(nextProps.max);
+        if (w != this.state.width) {
+            this.setState({width: this.calcWidth(nextProps.max)});
+        }
+    }
+    calcWidth(max) {
         return Math.floor(Math.log10(this.props.max)) + 1;
     }
     render() {
         return <div className={'spinbox' + (this.props.disabled ? ' disabled' : '')}>
             <input
                 type='number'
-                style={{width: this.calcWidth() + 'em'}}
+                style={{width: this.state.width + 'em'}}
                 min={this.props.min}
                 max={this.props.max}
                 value={this.props.value}
```

...?

Reply:
Yes. And probably should put inside a ```if(this.props.max!==nextProps.max)```

Reply:
> And probably should put inside a `if(this.props.max!==nextProps.max)`

Meh... compared to everything that already happens to cause `componentWillReceiveProps` to be called, I doubt the cost of calling `calcWidth` is noticeable.

Reply:
if(this.props.max!==nextProps.max){ calculateWidth() } would reduce the cost, since at least in thoery it's cheaper than the Math.log and Math.round.